designs/RestoreAndBuildCompatibility.md

```diff
@@ -0,0 +1,224 @@
+An example is the following Developer Community [issue](https://developercommunity2.visualstudio.com/t/NuGet-Restore-build-issues---projectass/1190427).
+
+#### Newer NuGet tooling, older build tooling
+
+Say a customer has configured to get latest NuGet in their build tasks. While NuGet.exe and VisualStudio do ship on the same day, the rollout to all build machines is not atomic. The customers could end up in this mismatched situation without an explicit action.
+
+A newer NuGet version *might* generate an assets file that has more/different details than what the build tasks expect.
+A recent example is the following Developer Community [issue](https://developercommunity2.visualstudio.com/t/vsbuild-msbuild-task-failing-since-2020-11-10/1249432) which is not 100% understood yet.
+
+In this design, we are trying to improve the general mismatched tooling experience, but also consider alternatives for the particular issues customers are hitting with VS 16.8/NuGet 5.8/ SDK 5.0.100.
+
+## Who are the customers
+
+PackageReference customers
+
+## Goals
+
+* Improve the experience for mismatched versions of restore/build tooling.
+
+## Non-Goals
+
+* Introducing explicit backwards/forward compatibility. That's a lateral effort.
+
+## Solution Overview
+
+In the 2 particular problems described above the experience when things go wrong is inconsistent. It's not *always* apparent why something goes wrong.
+This proposal covers improving that warning/error experience.
+
+### NuGet output versioning
+
+When PackageReference restore is run, NuGet writes out the 3 files mentioned above. 2 of those files have a `version`.
+
+* The `project.assets.json` has a version that has historically been used for `reading`, not tooling compatibility checks.
+* The `nuget.g.props` contains a `NuGetToolVersion` property.
+
+### Build Tasks warning (SDK)
+
+Given that the implied compatibility exists, the proposal is for the SDK build tasks raise a suppressible warning when the output is not in the expected shape.
+
+* Warn when the restore output is from an older tool than anticipated.
+* Warn when the restore output is from a newer tool than anticipated.
+
+There are 2 technical options and each might have a different effect on the frequency in which this warning is raised.
+
+#### Option 1 - Use assets file version
+
+NuGet increments the `project.assets.json` version every time there's a functional change in behavior. The build tasks in the SDK are built with a specific version of NuGet, and the SDK could raise a warning when the `project.assets.json` version is not the one it expects.
+
+Cons:
+
+* This version was not used with tooling/restore meaning compatibility in mind, but rather for writing/reading purposes. While unlikely, it's possible that other tooling depends on it.

```

Comment:
I don't know if this spec is designed to contain customer issues only. From an implementation point of view, it makes changes/insertions harder, because when we change NuGet and increment the assets file version, all the .NET SDK tests will fail, so we'll be blocked from insertions until the SDK is also updated to support it. Even when the assets file change is otherwise backwards compatible.

Reply:
It should be a suppressible warning raised at `build time`. 
It should only be an error if the reference resolution cannot succeed. 

Does that address your concern?
SDK side tests should be *easy* to fix, because they'll depend on `LockFileFormat.Version` or `LatestVersion`, https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.ProjectModel/LockFile/LockFileFormat.cs#L21, so it'll be `fixed` by building.